### PR TITLE
cluster provision with DBaaS

### PR DESCRIPTION
**Pull request by redhatHameed**
This PR enabled the provision of a crunchy bridge cluster with the DBaaS operator.  The changes have been done based on Provision ADR.

- Created CrunchyBridgeInstance controller, 
- updated DBaaSProviderSpec CR to include provision params and kind. 
-  Added SyncPeriod to reconciled the watched resource every 3 hours.
-  Added unit test for bridge instance controller.

### CODE REVIEW

**Comment by bmozaffa** [comment]
.DS_Store files are Mac generated... remove / ignore them?

**Comment by yzhao583** [comment]
@redhatHameed Do we have the link to redirect the users when they want to provision cluster for Crunchy?

**Comment by redhatHameed** [comment]
> @redhatHameed Do we have the link to redirect the users when they want to provision cluster for Crunchy?

added into provider registration CR [here](https://github.com/redhatHameed/crunchy-bridge-operator/blob/dbaas-provision/controllers/dbaas.redhat.com/dbaasprovider_reconciler.go#L64-L65)

File: controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+				logger.Info("cluster deleted", "id", id)
+			}
+
+			controllerutil.RemoveFinalizer(instanceObj, instanceFinalizer)
+			if err := r.Update(ctx, instanceObj); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		switch instanceObj.Status.Phase {
```
Comment (jianrongzhang89) [inline controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go:119]: If the user deletes the instance CR and subsequently (when the cluster is being deleted) the same CR is created, what's going to happen to the cluster? Will the deletion gets completed first and then the new cluster (with the same name) gets created?

File: controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+		}
+	}
+	return false
+}
+
+func (r *CrunchyBridgeInstanceReconciler) createFromSpec(spec dbaasv1alpha1.DBaaSInstanceSpec, bridgeapiClient *bridgeapi.Client) (bridgeapi.CreateRequest, error) {
+	var req bridgeapi.CreateRequest
+
+	req.Name = spec.Name
+	req.Provider = spec.CloudProvider
```
Comment (jianrongzhang89) [inline controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go:235]: If the provider or region is not provided by the user in the CR, should an error be raised here?

File: main.go [hunk 3]
```diff
@@ -84,6 +88,8 @@ func main() {
 	flag.BoolVar(&enableLeaderElection, "leader-elect", false,
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
+	flag.DurationVar(&syncPeriod, "sync-period", 3*time.Hour, "The minimum interval at which watched resources are reconciled (e.g. 12h)")
```
Comment (jianrongzhang89) [inline main.go:91]: Should sync-period be added to manager.yaml so that the value becomes configurable?

Comment (redhatHameed) [reply main.go]: added in env variable and command line

File: controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go (hunk 2 above)
Comment (redhatHameed) [reply controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go:235]: for crunchy suppose to provide these values, as no trial version option, we need to enable the validation in the first place-> CRD validation or webhook.

File: controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go (hunk 1 above)
Comment (redhatHameed) [reply controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go:119]: there is a finalizer on CR, so deletion performs first, then the user will be able to create the CR with the same name.

Comment (jianrongzhang89) [reply controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go:119]: What if the deletion operation in the finalizer fails or gets stuck? Will this cause the CR deletion to get stuck as well?

File: controllers/dbaas.redhat.com/conditions.go [hunk 4]
```diff
@@ -22,15 +22,17 @@ import (
...
-	Ready               string = "Ready"
-	NotFound            string = "NotFound"
-	SuccessMessage      string = "Successfully listed crunchy bridge Inventories"
-	SuccessConnection   string = "Successfully retrieved the connection detail\n"
+	SpecSynced             string = "SpecSynced"
+	BackendError           string = "BackendError"
+	AuthenticationError    string = "AuthenticationError"
+	SyncOK                 string = "SyncOK"
+	ReadyForBinding        string = "ReadyForBinding"
+	ClusterReady           string = "ClusterReady"
```
Comment (myeung18) [inline controllers/dbaas.redhat.com/conditions.go:30]: Is this Condition type? ADR is ```ProvisionReady```?
Do I missing anything? which one we should use?

Comment (redhatHameed) [reply controllers/dbaas.redhat.com/conditions.go:30]: Good catch @myeung18 its `ProvisionReady` I will update it.

File: controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go (hunk 1 above)
Comment (redhatHameed) [reply controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go:119]: that depends on the case, in which scenario cr get stuck, at the moment it works as expected.

File: controllers/dbaas.redhat.com/conditions.go (hunk 4 above)
Comment (redhatHameed) [reply controllers/dbaas.redhat.com/conditions.go]: updated

File: controllers/dbaas.redhat.com/inventory.go [hunk 5]
```diff
@@ -19,6 +19,7 @@ const (
 	CPU           = "cpu"
 	MEMORY        = "memory"
 	IS_HA         = "is_ha"
+	CLUSTER_NAME  = "name"
```
Comment (jianrongzhang89) [inline controllers/dbaas.redhat.com/inventory.go:22]: Should phase be added to the inventory instance info as per the ADR?

File: controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go (hunk 1 above)
Comment (jianrongzhang89) [reply controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go:119]: if bridgeapiClient.DeleteCluster() call (line #110) fails, the deletion may get stuck.

Comment (redhatHameed) [reply controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go:119]: so this expected behavior, if DeleteCluster fails then reconciling again, and it should not delete the CR until DeleteCluster return success.

File: controllers/dbaas.redhat.com/inventory.go (hunk 5 above)
Comment (redhatHameed) [reply controllers/dbaas.redhat.com/inventory.go:22]: need to update the ADR - see the current implementation : https://github.com/RHEcosystemAppEng/dbaas-operator/blob/4494a452c155eb4318ca4a854451f4de4c83da61/api/v1alpha1/dbaasprovider.go#L223

Comment (redhatHameed) [reply controllers/dbaas.redhat.com/inventory.go:22]: @jianrongzhang89 I see what you mean added the `state into instance info to get cluster state

File: config/samples/dbaas.redhat.com_v1alpha1_crunchybridgeinstance.yaml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    name: my-db-provider-account
+    namespace: openshift-dbaas-operator
+  name: malty-zebra-725
+  otherInstanceParams:
+    HighAvail: 'false'
+    PGMajorVer: '14'
+    Plan: hobby-2
+    Storage: '100'
+    TeamID: vp6hlxjcl5g73furjiztcrr2vi
+  provider: aws
```
Comment (wilybrace) [inline config/samples/dbaas.redhat.com_v1alpha1_crunchybridgeinstance.yaml:16]: update for cloudProvider/cloudRegion?

File: controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+limitations under the License.
+*/
+
+package dbaasredhatcom
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	crunchybridgev1alpha1 "github.com/CrunchyData/crunchy-bridge-operator/apis/crunchybridge/v1alpha1"
```
Comment (wilybrace) [inline controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go:23]: Please make/keep includes consistent with (core <line break> 3rd party <line break> org/local/project ) grouping used in the other go files. (one step in the right direction would use `goimports` on save over `gofmt`, as the former is a superset)

File: controllers/dbaas.redhat.com/crunchybridgeinstance_controller_test.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+import (
+	"github.com/CrunchyData/crunchy-bridge-operator/apis/dbaas.redhat.com/v1alpha1"
+	dbaasv1alpha1 "github.com/RHEcosystemAppEng/dbaas-operator/api/v1alpha1"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"time"
```
Comment (wilybrace) [inline controllers/dbaas.redhat.com/crunchybridgeinstance_controller_test.go:12]: Please make/keep includes consistent with (core <line break> 3rd party <line break> org/local/project ) grouping used in the other go files. (one step in the right direction would use `goimports` on save over `gofmt`, as the former is a superset)

File: controllers/dbaas.redhat.com/crunchybridgeinventory_controller.go [hunk 9]
```diff
@@ -18,26 +18,28 @@ package dbaasredhatcom
 
 import (
 	"context"
-	"net/url"
```
Comment (wilybrace) [inline controllers/dbaas.redhat.com/crunchybridgeinventory_controller.go:21]: Please make/keep includes consistent with (core <line break> 3rd party <line break> org/local/project ) grouping used in the other go files. (one step in the right direction would use `goimports` on save over `gofmt`, as the former is a superset)

File: controllers/dbaas.redhat.com/suite_test.go [hunk 10]
```diff
@@ -18,13 +18,14 @@ package dbaasredhatcom
 
 import (
 	"context"
+	"os"
```
Comment (wilybrace) [inline controllers/dbaas.redhat.com/suite_test.go:21]: Please make/keep includes consistent with (core <line break> 3rd party <line break> org/local/project ) grouping used in the other go files. (one step in the right direction would use `goimports` on save over `gofmt`, as the former is a superset)

File: main.go [hunk 11]
```diff
@@ -56,6 +58,7 @@ func init() {
 	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
 
 	utilruntime.Must(crunchybridgev1alpha1.AddToScheme(scheme))
+	utilruntime.Must(dbaasredhatcomv1alpha1.AddToScheme(scheme))
```
Comment (wilybrace) [inline main.go:61]: This belongs in `dbaas_support.go`, not `main.go`.

File: main.go [hunk 12]
```diff
@@ -39,6 +40,7 @@ import (
 	crunchybridgev1alpha1 "github.com/CrunchyData/crunchy-bridge-operator/apis/crunchybridge/v1alpha1"
 	crunchybridgecontrollers "github.com/CrunchyData/crunchy-bridge-operator/controllers/crunchybridge"
 
+	dbaasredhatcomv1alpha1 "github.com/CrunchyData/crunchy-bridge-operator/apis/dbaas.redhat.com/v1alpha1"
```
Comment (wilybrace) [inline main.go:43]: This inclusion a sure-fire signal that something is being added to `main.go` that should be in `dbaas_support.go`

**Review by wilybrace — CHANGES_REQUESTED** [review]
I'm not sure what else is pending from your other reviewers, but I would like to work toward getting this merged in so that the trial support can be added to it.

File: apis/crunchybridge/v1alpha1/bridgecluster_types.go [hunk 13]
```diff
@@ -22,9 +22,9 @@ import (
 
 const (
 	PhaseUnknown  = ""
-	PhasePending  = "pending"
-	PhaseCreating = "creating"
-	PhaseReady    = "ready"
+	PhasePending  = "Pending"
```
[the remark this answers is not in the document]
Comment (redhatHameed) [reply apis/crunchybridge/v1alpha1/bridgecluster_types.go:25]: Yes, for presentation purposes and following k8s standard the first letter Cap. WDYT this change will affect the logic of comparison, secondly all other providers operator following phase with the first letter in the capital?

File: bundle/manifests/crunchy-bridge-operator.clusterserviceversion.yaml [hunk 14]
```diff
@@ -43,6 +45,29 @@ metadata:
...
+            },
+            "name": "malty-zebra-725",
+            "otherInstanceParams": {
+              "HighAvail": "false",
+              "PGMajorVer": "14",
+              "Plan": "hobby-2",
+              "Storage": "100",
+              "TeamID": "vp6hlxjcl5g73furjiztcrr2vi"
+            },
+            "provider": "aws",
```
[the remark this answers is not in the document]
Comment (redhatHameed) [reply bundle/manifests/crunchy-bridge-operator.clusterserviceversion.yaml]: Agreed changed accordingly

File: config/samples/dbaas.redhat.com_v1alpha1_crunchybridgeinstance.yaml (hunk 6 above)
Comment (redhatHameed) [reply config/samples/dbaas.redhat.com_v1alpha1_crunchybridgeinstance.yaml]: updated

File: controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go (hunk 7 above)
Comment (redhatHameed) [reply controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go]: used goimprot to group the packages.

File: controllers/dbaas.redhat.com/crunchybridgeinstance_controller_test.go (hunk 8 above)
Comment (redhatHameed) [reply controllers/dbaas.redhat.com/crunchybridgeinstance_controller_test.go:13]: used goimprot to group the packages.

File: main.go (hunk 11 above)
Comment (redhatHameed) [reply main.go]: moved into dbaas_support.go

**Comment by redhatHameed** [comment]
@wilybrace is there any other comments need to address or have any concern to merge this PR

**Comment by wilybrace** [comment]
will address any lingering with my changset.

**Pull request merged**